File: react-native/storyly-react-native/ios/RCTConvert+Storyly.m
//
//  RCTConvert+StorylyInit.m
//  storyly-react-native
//
//  Created by Levent Oral on 12.05.2020.
//

#import "RCTConvert+Storyly.h"

@implementation RCTConvert (StorylyInit)

+ (StorylyInit *)STStorylyInit:(id)json {
    NSDictionary *storylyInitJson = [self NSDictionary:json];
    StorylySegmentation *storylySegmentation = [StorylySegmentation alloc];
    if ([storylyInitJson.allKeys containsObject:@"storylySegments"] &&
        storylyInitJson[@"storylySegments"] != NULL) {
        storylySegmentation = [storylySegmentation initWithSegments:storylyInitJson[@"storylySegments"]];
    }
    
    bool isTestMode = ([storylyInitJson.allKeys containsObject:@"storylyIsTestMode"]) ? [storylyInitJson[@"storylyIsTestMode"] boolValue] : NO;

    StorylyInit *storylyInit = [[StorylyInit alloc] initWithStorylyId:storylyInitJson[@"storylyId"]
                                                         segmentation:storylySegmentation
                                                      customParameter:storylyInitJson[@"customParameter"]
                                                           isTestMode:isTestMode];
    [storylyInit setUserData:storylyInitJson[@"userProperty"]];
    return storylyInit;
}

@end

@implementation RCTConvert (StoryGroupListStyling)

+ (StoryGroupListStyling *)STStoryGroupListStyling:(id)json {
    NSDictionary *storyGroupListStyling = [self NSDictionary:json];
    
    float edgePadding = 4;
    if ([storyGroupListStyling.allKeys containsObject:@"edgePadding"] && storyGroupListStyling[@"edgePadding"] != NULL) {
        edgePadding = [storyGroupListStyling[@"edgePadding"] floatValue];
    }
    
    float paddingBetweenItems = 8;
    if ([storyGroupListStyling.allKeys containsObject:@"paddingBetweenItems"] && storyGroupListStyling[@"paddingBetweenItems"] != NULL) {
        paddingBetweenItems = [storyGroupListStyling[@"paddingBetweenItems"] floatValue];
    }

    return [[StoryGroupListStyling alloc] initWithEdgePadding:edgePadding
                                          paddingBetweenItems:paddingBetweenItems];
}

@end

@implementation RCTConvert (StoryGroupIconStyling)

+ (StoryGroupIconStyling *)STStoryGroupIconStyling:(id)json {
    NSDictionary *storyGroupIconStyling = [self NSDictionary:json];
    
    float height = 80;
    if ([storyGroupIconStyling.allKeys containsObject:@"height"] && storyGroupIconStyling[@"height"] != NULL) {
        height = [storyGroupIconStyling[@"height"] floatValue];
    }
    
    float width = 80;
    if ([storyGroupIconStyling.allKeys containsObject:@"width"] && storyGroupIconStyling[@"width"] != NULL) {
        width = [storyGroupIconStyling[@"width"] floatValue];
    }
    
    float cornerRadius = 40;
    if ([storyGroupIconStyling.allKeys containsObject:@"cornerRadius"] && storyGroupIconStyling[@"cornerRadius"] != NULL) {
        cornerRadius = [storyGroupIconStyling[@"cornerRadius"] floatValue];
    }
    

    return [[StoryGroupIconStyling alloc] initWithHeight:height
                                                   width:width
                                            cornerRadius:cornerRadius];
}

@end

@implementation RCTConvert (StoryGroupTextStyling)

+ (StoryGroupTextStyling *)STStoryGroupTextStyling:(id)json {
    NSDictionary *storyGroupTextStyling = [self NSDictionary:json];
    
    BOOL isVisible = YES;
    if ([storyGroupTextStyling.allKeys containsObject:@"isVisible"] && storyGroupTextStyling[@"isVisible"] != NULL) {
        isVisible = [storyGroupTextStyling[@"isVisible"] boolValue];
    }
    
    
    UIColor* textColor = UIColor.blackColor;
    if ([storyGroupTextStyling.allKeys containsObject:@"color"] && storyGroupTextStyling[@"color"] != NULL) {
        textColor = [self getUIColorObjectFromHexString:storyGroupTextStyling[@"color"]];
    }

    UIColor textColorSeen = NULL;
    if ([storyGroupTextStyling.allKeys containsObject:@"colorSeen"] && storyGroupTextStyling[@"colorSeen"] != NULL) {
        textColorSeen = [self getUIColorObjectFromHexString:storyGroupTextStyling[@"colorSeen"]];
    }

    UIColor textColorSeen = NULL;
    if ([storyGroupTextStyling.allKeys containsObject:@"colorUnseen"] && storyGroupTextStyling[@"colorUnseen"] != NULL) {
        textColorSeen = [self getUIColorObjectFromHexString:storyGroupTextStyling[@"colorUnseen"]];
    }
    
    int fontSize = 12;
    if ([storyGroupTextStyling.allKeys containsObject:@"textSize"] && storyGroupTextStyling[@"textSize"] != NULL) {
        fontSize = [storyGroupTextStyling[@"textSize"] intValue];
    }
    
    int lines = 2;
    if ([storyGroupTextStyling.allKeys containsObject:@"lines"] && storyGroupTextStyling[@"lines"] != NULL) {
        lines = [storyGroupTextStyling[@"lines"] intValue];
    }
    
    return [[StoryGroupTextStyling alloc] initWithIsVisible:isVisible
                                                      colorSeen:colorSeen
                                                      colorUnseen:colorUnseen
                                                      font:[UIFont systemFontOfSize:fontSize]
                                                      lines:lines];
}

+ (UIColor *)getUIColorObjectFromHexString:(NSString *)hexStr {
    unsigned int hexInt = 0;
    NSString *noHashString = [hexStr stringByReplacingOccurrencesOfString:@"#" withString:@""];
    NSScanner *scanner = [NSScanner scannerWithString:noHashString];
    [scanner setCharactersToBeSkipped:[NSCharacterSet symbolCharacterSet]];
    [scanner scanHexInt:&hexInt];
    
    CGFloat alpha, red, green, blue;
    if ([hexStr length] == 9) {
        alpha = ((hexInt & 0xFF000000) >> 24) / 255.0f;
        red = ((hexInt & 0x00FF0000) >> 16) / 255.0f;
        green = ((hexInt & 0x0000FF00) >> 8) / 255.0f;
        blue = (hexInt & 0x000000FF) / 255.0f;
    } else {
        alpha = 1;
        red = ((hexInt & 0xFF0000) >> 16) / 255.0f;
        green = ((hexInt & 0x00FF00) >>  8) / 255.0f;
        blue = (hexInt & 0x0000FF) / 255.0f;
    }
    return [[UIColor alloc] initWithRed:red
                                  green:green
                                   blue:blue
                                  alpha:alpha];
}

@end

@implementation RCTConvert (StoryHeaderStyling)

+ (StoryHeaderStyling *)STStoryHeaderStyling:(id)json {
    NSDictionary *storyHeaderStyling = [self NSDictionary:json];
    
    BOOL isTextVisible = YES;
    if ([storyHeaderStyling.allKeys containsObject:@"isTextVisible"] && storyHeaderStyling[@"isTextVisible"] != NULL) {
        isTextVisible = [storyHeaderStyling[@"isTextVisible"] boolValue];
    }
    
    BOOL isIconVisible = YES;
    if ([storyHeaderStyling.allKeys containsObject:@"isIconVisible"] && storyHeaderStyling[@"isIconVisible"] != NULL) {
        isIconVisible = [storyHeaderStyling[@"isIconVisible"] boolValue];
    }
    
    BOOL isCloseButtonVisible = YES;
    if ([storyHeaderStyling.allKeys containsObject:@"isCloseButtonVisible"] && storyHeaderStyling[@"isCloseButtonVisible"] != NULL) {
        isCloseButtonVisible = [storyHeaderStyling[@"isCloseButtonVisible"] boolValue];
    }

    return [[StoryHeaderStyling alloc] initWithIsTextVisible:isTextVisible
                                               isIconVisible:isIconVisible
                                        isCloseButtonVisible:isCloseButtonVisible
                                             closeButtonIcon:NULL
                                             shareButtonIcon:NULL];
}

@end

File: react-native/storyly-react-native/ios/RCTConvert+Storyly.swift
//
//  RCTConvert+Storyly.swift
//  storyly-react-native
//
//  Created by Haldun Melih Fadillioglu on 25.10.2022.
//
import Storyly

@objc(StorylyBundle)
class StorylyBundle: NSObject {
    let storylyView: StorylyView
    let storyGroupViewFactory: STStoryGroupViewFactory?
    
    init(storylyView: StorylyView,
         storyGroupViewFactory: STStoryGroupViewFactory?) {
        self.storylyView = storylyView
        self.storyGroupViewFactory = storyGroupViewFactory
    }
}

extension RCTConvert {
    @objc(stStorylyBundle:)
    static func stStorylyBundle(json: NSDictionary) -> StorylyBundle? {
        print("STR:RCTConvert+Extension:stStorylyBundle(json:\(json))")
        guard let storylyInitJson = json["storylyInit"] as? NSDictionary else { return nil }
        guard let storyGroupIconStylingJson = json["storyGroupIconStyling"] as? NSDictionary else { return nil }
        guard let storyGroupViewFactoryJson = json["storyGroupViewFactory"] as? NSDictionary else { return nil }
        guard let storyGroupListStylingJson = json["storyGroupListStyling"] as? NSDictionary else { return nil }
        guard let storyGroupTextStylingJson = json["storyGroupTextStyling"] as? NSDictionary else { return nil }
        guard let storyHeaderStylingJson = json["storyHeaderStyling"] as? NSDictionary else { return nil }
        
        var storyGroupViewFactory: STStoryGroupViewFactory? = nil
        
        let storylyView = StorylyView()
        storylyView.storylyInit = stStorylyInit(json: storylyInitJson)
        storylyView.storylyShareUrl = json["storylyShareUrl"] as? String

        if let storyGroupViewFactorySize = stStoryGroupViewFactorySize(json: storyGroupViewFactoryJson) {
            storyGroupViewFactory = STStoryGroupViewFactory(width: storyGroupViewFactorySize.width,
                                                            height: storyGroupViewFactorySize.height)
        } else {
            storylyView.storyGroupSize = json["storyGroupSize"] as? String ?? "large"
            storylyView.storyGroupIconStyling = stStoryGroupIconStyling(json: storyGroupIconStylingJson)
            storylyView.storyGroupTextStyling = stStoryGroupTextStyling(json: storyGroupTextStylingJson)
            if let storyGroupIconBorderColorSeenJson = json["storyGroupIconBorderColorSeen"] as? NSArray {
                storylyView.storyGroupIconBorderColorSeen = RCTConvert.uiColorArray(storyGroupIconBorderColorSeenJson)
            }
            if let storyGroupIconBorderColorNotSeenJson = json["storyGroupIconBorderColorNotSeen"] as? NSArray {
                storylyView.storyGroupIconBorderColorNotSeen = RCTConvert.uiColorArray(storyGroupIconBorderColorNotSeenJson)
            }
            if let storyGroupIconBackgroundColorJson = json["storyGroupIconBackgroundColor"] as? NSNumber {
                storylyView.storyGroupIconBackgroundColor = RCTConvert.uiColor(storyGroupIconBackgroundColorJson)
            }
            if let storyGroupPinIconColorJson = json["storyGroupPinIconColor"] as? NSNumber {
                storylyView.storyGroupPinIconColor = RCTConvert.uiColor(storyGroupPinIconColorJson)
            }
            storylyView.storyGroupAnimation = json["storyGroupAnimation"] as? String ?? "border-rotation"
        }
        storylyView.storyGroupListStyling = stStoryGroupListStyling(json: storyGroupListStylingJson)
        if stStorylyLayoutDirection(direction: (json["storylyLayoutDirection"] as? String) ?? "ltr") == .RTL {
            storylyView.storylyLayoutDirection = .RTL
        }
        
        storylyView.storyHeaderStyling = stStoryHeaderStyling(json: storyHeaderStylingJson)
        if let storyItemTextColorJson = json["storyItemTextColor"] as? NSNumber {
            storylyView.storyItemTextColor = RCTConvert.uiColor(storyItemTextColorJson)
        }
        if let storyItemIconBorderColorJson = json["storyItemIconBorderColor"] as? NSNumber {
            storylyView.storyItemIconBorderColor = RCTConvert.uiColorArray(storyItemIconBorderColorJson)
        }
        if let storyItemProgressBarColorJson = json["storyItemProgressBarColor"] as? NSNumber {
            storylyView.storylyItemProgressBarColor = RCTConvert.uiColorArray(storyItemProgressBarColorJson)
        }
        storylyView.storyItemTextFont = getCustomFont(typeface: json["storyItemTextTypeface"] as? NSString, fontSize: 14, defaultWeight: .semibold)
        storylyView.storyInteractiveFont = getCustomFont(typeface: json["storyInteractiveTextTypeface"] as? NSString, fontSize: 14, defaultWeight: .regular)
        
        return StorylyBundle(storylyView: storylyView, storyGroupViewFactory: storyGroupViewFactory)
    }
}

private func stStorylyInit(json: NSDictionary) -> StorylyInit {
    var segmentation: StorylySegmentation = StorylySegmentation(segments: nil)
    if let segmentsData = json["storylySegments"] as? [String] {
        segmentation = StorylySegmentation(segments: Set(segmentsData))
    }
    
    return StorylyInit(
        storylyId: (json["storylyId"] as? String) ?? "",
        segmentation: segmentation,
        customParameter: json["customParameter"] as? String,
        isTestMode: (json["storylyIsTestMode"] as? Bool) ?? false,
        storylyPayload: json["storylyPayload"] as? String,
        userData: json["userProperty"] as? [String: String] ?? [:]
    )
}

private func stStoryGroupListStyling(json: NSDictionary) -> StoryGroupListStyling {
    var orientation: StoryGroupListOrientation
    switch json["orientation"] as? String {
        case "horizontal": orientation = .Horizontal
        case "vertical": orientation = .Vertical
        default: orientation = .Horizontal
    }

    let sections = (json["sections"] as? Int) ?? 1
    let horizontalEdgePadding = (json["horizontalEdgePadding"] as? CGFloat) ?? 4
    let verticalEdgePadding = (json["verticalEdgePadding"] as? CGFloat) ?? 4
    let horizontalPaddingBetweenItems = (json["horizontalPaddingBetweenItems"] as? CGFloat) ?? 8
    let verticalPaddingBetweenItems = (json["verticalPaddingBetweenItems"] as? CGFloat) ?? 8

    let storylyGroupListStyling = StoryGroupListStyling(
        orientation: orientation,
        sections: sections,
        horizontalEdgePadding: horizontalEdgePadding,
        verticalEdgePadding: verticalEdgePadding,
        horizontalPaddingBetweenItems: horizontalPaddingBetweenItems,
        verticalPaddingBetweenItems: verticalPaddingBetweenItems
    )
    return storylyGroupListStyling
}

private func stStoryGroupIconStyling(json: NSDictionary) -> StoryGroupIconStyling {
    let height = (json["height"] as? CGFloat) ?? 80
    let width = (json["width"] as? CGFloat) ?? 80
    let cornerRadius = (json["cornerRadius"] as? CGFloat) ?? 40
    return StoryGroupIconStyling(height: height, width: width, cornerRadius: cornerRadius)
}

private func stStoryGroupViewFactorySize(json: NSDictionary) -> CGSize? {
    let width = (json["width"] as? CGFloat) ?? 0
    let height = (json["height"] as? CGFloat) ?? 0
    let factorySize = CGSize(width: width, height: height)
    return factorySize == .zero ? nil : factorySize
}

private func stStoryGroupTextStyling(json: NSDictionary) -> StoryGroupTextStyling {
    let isVisible = (json["isVisible"] as? Bool) ?? true
    let textColorSeen = UIColor(hexString: (json["colorSeen"] as? String)) ?? .black
    let textColorNotSeen = UIColor(hexString: (json["colorNotSeen"] as? String)) ?? .black
    
    let fontSize = (json["textSize"] as? Int) ?? 12
    let font = getCustomFont(typeface: (json["typeface"] as? NSString), fontSize: CGFloat(fontSize))
    let lines = (json["lines"] as? Int) ?? 2
    
    return StoryGroupTextStyling(isVisible: isVisible,
                                 colorSeen: textColorSeen,
                                 colorNotSeen: textColorNotSeen,
                                 font: font,
                                 lines: lines)
}

private func stStorylyLayoutDirection(direction: String) -> StorylyLayoutDirection {
    switch direction {
        case "ltr": return .LTR
        case "rtl": return .RTL
        default: return .LTR
    }
}

private func stStoryHeaderStyling(json: NSDictionary) -> StoryHeaderStyling {
    let isTextVisible = (json["isTextVisible"] as? Bool) ?? true
    let isIconVisible = (json["isIconVisible"] as? Bool) ?? true
    let isCloseButtonVisible = (json["isCloseButtonVisible"] as? Bool) ?? true
    
    let closeIconImage = UIImage(named: (json["closeIcon"] as? String))
    let shareIconImage = UIImage(named: (json["shareIcon"] as? String))
    
    return StoryHeaderStyling(isTextVisible: isTextVisible,
                              isIconVisible: isIconVisible,
                              isCloseButtonVisible: isCloseButtonVisible,
                              closeButtonIcon: closeIconImage,
                              shareButtonIcon: shareIconImage)
}

private func getCustomFont(typeface: NSString?, fontSize: CGFloat, defaultWeight: UIFont.Weight = .regular) -> UIFont {
    if let fontName = typeface?.deletingPathExtension,
       let font = UIFont(name: fontName, size: fontSize) { return font }
    return .systemFont(ofSize: fontSize, weight: defaultWeight)
}

internal extension UIImage {
    convenience init?(named: String?) {
        guard let named = named else { return nil }
        self.init(named: named)
    }
}

internal extension UIColor {
    convenience init?(hexString: String?) {
        guard let hexString = hexString else { return nil }

        let start = hexString.index(hexString.startIndex, offsetBy: 1)
        let hexColor = String(hexString[start...])
        
        let scanner = Scanner(string: hexColor)
        var hexNumber: UInt64 = 1
        scanner.scanHexInt64(&hexNumber)
        
        let red, green, blue, alpha: CGFloat
        if hexString.count == 9 {
            alpha = CGFloat((hexNumber & 0xff000000) >> 24) / 255
            red = CGFloat((hexNumber & 0x00ff0000) >> 16) / 255
            green = CGFloat((hexNumber & 0x0000ff00) >> 8) / 255
            blue = CGFloat(hexNumber & 0x000000ff) / 255
        } else {
            red = CGFloat((hexNumber & 0xff0000) >> 16) / 255
            green = CGFloat((hexNumber & 0x00ff00) >> 8) / 255
            blue = CGFloat(hexNumber & 0x0000ff) / 255
            alpha = 1
        }
        self.init(red: red, green: green, blue: blue, alpha: alpha)
    }
}
